packages/std/src/ibc/transfer_msg_builder.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{
+    to_json_string, Coin, IbcCallbackRequest, IbcDstCallback, IbcMsg, IbcSrcCallback, IbcTimeout,
+};
+
+// these are the different memo types and at the same time the states
+// the TransferMsgBuilder can be in
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct EmptyMemo;
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WithMemo {
+    memo: String,
+}
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WithSrcCallback {
+    src_callback: IbcSrcCallback,
+}
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WithDstCallback {
+    dst_callback: IbcDstCallback,
+}
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WithCallbacks {
+    src_callback: IbcSrcCallback,
+    dst_callback: IbcDstCallback,
+}
+
+impl From<EmptyMemo> for Option<String> {
+    fn from(_: EmptyMemo) -> Self {
+        None
+    }
+}
+
+impl From<WithMemo> for Option<String> {
+    fn from(m: WithMemo) -> Self {
+        Some(m.memo)
+    }
+}
+
+impl From<WithSrcCallback> for Option<String> {
+    fn from(s: WithSrcCallback) -> Self {
+        Some(to_json_string(&IbcCallbackRequest::source(s.src_callback)).unwrap())
+    }
+}
+
+impl From<WithDstCallback> for Option<String> {
+    fn from(d: WithDstCallback) -> Self {
+        Some(to_json_string(&IbcCallbackRequest::destination(d.dst_callback)).unwrap())
+    }
+}
+
+impl From<WithCallbacks> for Option<String> {
+    fn from(c: WithCallbacks) -> Self {
+        Some(to_json_string(&IbcCallbackRequest::both(c.src_callback, c.dst_callback)).unwrap())
+    }
+}
+
+impl<T: Into<Option<String>>> TransferMsgBuilder<T> {
+    pub fn build(self) -> IbcMsg {
+        IbcMsg::Transfer {
+            channel_id: self.channel_id,
+            to_address: self.to_address,
+            amount: self.amount,
+            timeout: self.timeout,
+            memo: self.memo.into(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TransferMsgBuilder<MemoData> {
+    channel_id: String,
+    to_address: String,
+    amount: Coin,
+    timeout: IbcTimeout,
+    memo: MemoData,
```

Comment:
Isn't this generic type overkill? As far as I can see, this is PR is only used for type-safe construction, not usage of the typed builder data.

I think there are ways to simplify:
1. Make `memo: ` an enum
2. Make `memo: ` an `Option<String>` and serialize in the constructor

Reply:
Making memo an `Option<String>` doesn't play well with having chained `.with_src_callback(src_callback).with_dst_callback(dst_callback)`.
An enum for memo works better, but loses the ability to prevent you from calling `.with_src_callback(src_callback).with_memo(memo)`, which would overwrite your src callback without it being obvious.

Reply:
Okay, now I get it. Thanks for the explanation. Let's keep it as-is.

However, what about an explicit trait to improve clarity of the code. E.g.

```rust
trait MemoSource {
    // we might need to make this failable at some point further down the road
    fn into_memo() -> Option<String>;
}
```

Reply:
Yes, I think that makes sense